### PR TITLE
arroyo router prototype and stale message rerouting

### PR DESCRIPTION
**Pull request by lynnagara**
## this pr does 3 main things:
1. adds a router component
2. adds a guard component
3. routes stale transactions to an alternate topic if cli arg is passed

## 1. router component

the router strategy enables forking of any strategy and merging back together whilst maintaining original message ordering

this is inspired by https://github.com/getsentry/arroyo/pull/368/ but has a few differences:

- it can be used flexibly before or after any strategy. multiple routers can also be chained together. there is no requirement for the router to be the last step in the chain.

- it offloads handling of backpressure and invalid messages to the guard component

- it offloads handling of commit logic to CommitOffsets, which is not part of the router

- it maintains original message order to avoid additional watermark math

## 2. guard component

the goal of the guard component is to handle backpressure for strategies in a generic way. this logic is easy to write wrong and the interactions between components can be hard to test.

this is mostly designed so that it can be used as-is by arroyo's existing strategies

## 3. transactions implementation

the transactions consumer now supports a cli arg `--stale-threshold-sec`. if passed, messages older than this threshold will be rerouted to a secondary topic (currently dlq but can be changed when one is provisioned).

the default behavior (no rerouting) is unchanged

### CODE REVIEW

File: src/sentry/utils/arroyo_guard.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    [type[NonFinalStrategy[TStrategyPayload, TResult]]],
+    type[NonFinalStrategy[TStrategyPayload, TResult]],
+]:
+    """
+    Handles backpressure and invalid messages for any strategy. If
+    MessageRejected or InvalidMessage is received from the inner strategy
+    from the inner strategy the message is carried over by the guard for
+    future submission / reraise so the parent strategy does not have to implement
+    any handling for this scenario and can safely call `next_step.submit()`
+    in any part of the strategy.
```
Comment (untitaker) [inline src/sentry/utils/arroyo_guard.py:40]: this i don't get. if the buffer reaches `max_buffer_size`, `submit` can still raise `MessageRejected`. you're not adding any guarantees with this buffer. the backpressure has to go somewhere: either it is propagated all the time or we OOM

Comment (lynnagara) [reply src/sentry/utils/arroyo_guard.py:40]: there's a couple of scenarios that come to mind:

- inner submit called outside the parent's submit method:. for example in https://github.com/getsentry/arroyo/blob/main/arroyo/processing/strategies/produce.py#L85 we raise MessageRejected on poll even though we are not supposed to
- a message passed to a parent strategy might be unbatched into more than one message to be forwarded to the next step downsteram. If one message in the middle fails with messagerejected/invalid message we need to resubmit only the second half of the batch not the whole thing. this is kinda the same thing we do in unfold today (https://github.com/getsentry/arroyo/blob/main/arroyo/processing/strategies/unfold.py#L37) where we put the remaining ones into a buffer, but this is an attempt to make that idea more reusable

File: src/sentry/utils/arroyo_guard.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+                    if isinstance(self.__messages_carried_over[0], InvalidMessage):
+                        raise self.__messages_carried_over.popleft()
+
+                    try:
+                        self.__inner_submit(self.__messages_carried_over[0])
+                        self.__messages_carried_over.popleft()
+                    except MessageRejected:
+                        break
+                    except InvalidMessage as exc:
+                        self.__messages_carried_over.append(exc)
```
Comment (untitaker) [inline src/sentry/utils/arroyo_guard.py:86]: if we `append` `InvalidMessage` to the right side of the queue, and then pop from the left, we might delay the re-raising of `InvalidMessage` for a while. I'm not sure but I think this means we might re-raise `InvalidMessage` when the referred message has already been committed, at which point we can't DLQ it anymore. Is there a reason to not let the exception bubble up directly?

File: src/sentry/utils/arroyo_router.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+            route.join(remaining)
+
+        remaining = deadline - time.time() if deadline else None
+        self.next_step.join(deadline)
+
+    def close(self) -> None:
+        pass
+
+    def terminate(self) -> None:
+        pass
```
Comment (untitaker) [inline src/sentry/utils/arroyo_router.py:93]: close/terminate all routes?

File: src/sentry/utils/arroyo_guard.py (hunk 1 above)
Comment (untitaker) [reply src/sentry/utils/arroyo_guard.py:40]: ah I misread the code. i think the intention makes sense.

File: src/sentry/utils/arroyo_guard.py (hunk 2 above)
Comment (lynnagara) [reply src/sentry/utils/arroyo_guard.py:86]: an invalid message from downstream by definition must be earlier offset than a message that hasn't even been submitted yet, so technically it's always safe to reraise those first
though putting it in a buffer rather than straight reraising is supposed to solve this problem:
```
for msg in messages_for_next_step:
   inner_step.submit(msg)
```
and potentially getting an invalid message from some of these calls

Comment (untitaker) [reply src/sentry/utils/arroyo_guard.py:86]: > so technically it's always safe to reraise those first

but you aren't. your queue has offsets like this (as in `wrapped_inner_submit` you append to the right):

```
1, 2, 3, 4, 5
```

you use `popleft`, so you submit 1, 2 to next_step, and next_step raises that 2 is invalid. so now you `append` on the right:

```
3, 4, 5, invalid(2)
```

on the next poll, you forward 3, 4, 5 to next_step, and next_step commits 5.

you then raise InvalidMessage(2), but 5 has already been committed.

tldr I think you want `appendleft` here, but i'm not confident that'll be sufficient

File: src/sentry/utils/arroyo_guard.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    `next_step` must be the first arg in the constructor.
+
+    """
+
+    def guard_inner(
+        cls: type[NonFinalStrategy[TStrategyPayload, TResult]]
+    ) -> type[NonFinalStrategy[TStrategyPayload, TResult]]:
+
+        class Guard(NonFinalStrategy[TStrategyPayload, TResult]):
+            def __init__(
```
Comment (untitaker) [inline src/sentry/utils/arroyo_guard.py:51]: this is a bit confusing, and I had to do a double-take to see what is being wrapped. since it appears to only wrap next_step, I think it can be modelled as a strategy around next_step. instead of:

```
@guard
class StepOne(...): ...
```

one could write:

```
class StepOne(...):
    def __init__(self, next_step):
        self.next_step = Guard(next_step)
```

the code in `poll()` can live in `Guard.poll`, as `StepOne` will call `Guard.poll()` anyway.

File: src/sentry/utils/arroyo_guard.py (hunk 2 above)
Comment (lynnagara) [reply src/sentry/utils/arroyo_guard.py:86]: sorry yeah i know this isn't happening now - i think the most correct way to achieve this is to actually maintain a separate queue for the invalid message exceptions so they are also in proper order and ensure that's empty first before processing anything in `self.__messages_carried_over`. i can add that if doing it that way makes sense to you?

Comment (untitaker) [reply src/sentry/utils/arroyo_guard.py:86]: i think that makes sense

File: src/sentry/utils/arroyo_guard.py (hunk 4 above)
Comment (lynnagara) [reply src/sentry/utils/arroyo_guard.py:51]: oh.. this sounds way simpler!

File: src/sentry/utils/arroyo_guard.py [hunk 5]
```diff
@@ -73,10 +74,10 @@ def submit(self, msg: Message[TStrategyPayload]) -> None:
                 self.__instance.submit(msg)
 
             def poll(self) -> None:
-                while self.__messages_carried_over:
-                    if isinstance(self.__messages_carried_over[0], InvalidMessage):
-                        raise self.__messages_carried_over.popleft()
+                while self.__invalid_messages:
+                    raise self.__messages_carried_over.popleft()
```
Comment (untitaker) [inline src/sentry/utils/arroyo_guard.py:78]:
```suggestion
                    raise self.__invalid_messages.popleft()
```

File: src/sentry/utils/arroyo_guard.py (hunk 4 above)
Comment (lynnagara) [reply src/sentry/utils/arroyo_guard.py:51]: ok there's one thing i can't figure out how to deal with if we move this way

StepOne doesn't know when the guard's buffer is full, which is needed to raise MessageRejected to the parent

In the current implementation the guard controls both the parent and inner strategy and can do this

Do you have any ideas to deal with this?

File: src/sentry/utils/arroyo_router.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+            for (routing_key, builder) in route_builders.items()
+        }
+
+    def submit(self, message: Message[TStrategyPayload]) -> None:
+        if len(self.buffer) > 10000:
+            raise MessageRejected
+
+        message_route = self.routing_func(message)
+        self.routes[message_route].submit(message)
+        self.buffer.remove(message, message_route)
```
Comment (fpacifici) [inline src/sentry/utils/arroyo_router.py:65]: I am not sure I follow this process in the following cases:
- Messages filtered by one of the routes. We would try to remove a message that potentially was never added to the buffer. Is that ok ? 
- Batching messages. Let's say one route is baching messages and the other is not. If we send 10 messages to the batching strategy then a message to the non batching one then 10 messages to the batching strategy, we may remove the individual message before the batch is even added. Is this a problem ?

File: src/sentry/utils/arroyo_router.py [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+    def add(self, message: Message[TResult], routing_key: str) -> None:
+        self.messages.append((routing_key, message))
+
+    def remove(self, message: Message[Any], routing_key: str) -> None:
+        for partition, committable_offset in message.committable.items():
+
+            if self.committed_offsets[routing_key].get(partition):
+                if committable_offset > self.committed_offsets[routing_key][partition]:
+                    self.committed_offsets[routing_key][partition] = committable_offset
```
Comment (fpacifici) [inline src/sentry/utils/arroyo_router.py:122]: Does `self.committed_offsets` refer to actually committed offsets in Kafka? It does not seems from the implementation as we remove a message from the buffer right after submitting.

File: src/sentry/utils/arroyo_router.py [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+            raise MessageRejected
+
+        message_route = self.routing_func(message)
+        self.routes[message_route].submit(message)
+        self.buffer.remove(message, message_route)
+
+        while True:
+            next_message = self.buffer.poll()
+            if next_message:
+                self.next_step.submit(next_message)  # this is fine because of backpresure_guard
```
[the remark this answers is not in the document]
Comment (lynnagara) [reply src/sentry/utils/arroyo_router.py:70]: - currently the messages are just in the order they finished processing by the routes, and are not reordered by offsets.
- guaranteeing ordering by offsets is impossible because you can always have one route with `{partition1: offset2, partition2: offset1}` and another route with `{partition1: offset1, partition2:offset2}` and there is no valid total order
- so instead all we do is hold onto the message until it's committable offset is below the last processed one on all routes
- so the batch can occur before or after the individual message but we only process it after the individual message has definitely finished processing, and vice versa

File: src/sentry/utils/arroyo_router.py (hunk 7 above)
Comment (lynnagara) [reply src/sentry/utils/arroyo_router.py]: sorry, this was supposed to be `committable_offsets` not `committed_offsets`

File: src/sentry/utils/arroyo_router.py (hunk 6 above)
Comment (lynnagara) [reply src/sentry/utils/arroyo_router.py:65]: we should probably change the remove api but it does nothing with the message itself, so filtered or batched or whatever doesn't matter. it just marks those offsets as complete for that route.

**Comment by natemoo-re** [comment]
Hey! We are cleaning up our PR backlog to cut down on noise. Since this draft was created more than 6 months ago and doesn't have much recent activity, I'm closing it.

If you’re still interested in seeing this land, please feel free to reopen.
